aip/0235.md

```diff
@@ -0,0 +1,86 @@
+---
+aip:
+  id: 235
+  state: draft
+  created: 2019-06-18
+permalink: /235
+redirect_from:
+  - /0235
+---
+
+# Batch Delete
+
+Some APIs need to allow users to delete a set of resources in a single transaction. A batch delete method provides this functionality.
+
+## Guidance
+
+Batch delete methods are specified using the following pattern:
+
+```proto
+rpc BatchDeleteBooks(BatchDeleteBooksRequest) returns (google.protobuf.Empty) {
+  option (google.api.http) = {
+    post: "/v1/shelves/-/books:batchDelete"
+    body: "*"
+  };
+}
+```
+
+- The RPC's name **must** begin with `BatchDelete`. The remainder of the RPC name
+  **should** be the plural form of the resource being deleted.
+- The request message **must** match the RPC name, with a `-Request` suffix.
+- The response message **should** be `google.protobuf.Empty`.
+  - If the resource is [soft deleted](#soft-delete), the response message
+    **should** be a response message containing the updated resources.
+  - In the event that the request may take a significant amount of time, the
+    response message **must** be a `google.longrunning.Operation` which resolves to
+    the correct response.
+- The HTTP verb **must** be `POST` (not `DELETE`).
+- The HTTP URI **must** end with `:batchDelete`.
+- The URL path **should** represent the collection for the resource, matching
+  the collection used for simple CRUD operations. If the operation spans
+  parents, a dash (`-`) **may** be used as a wildcard.
+- The body clause in the `google.api.http` annotation **should** be `"*"`.
+- The operation **should** be atomic: it **should** fail for all resources or succeed
+  for all resources (no partial success).
+
+### Request message
+
+The request for a batch delete method **should** be specified with the following
+pattern:
+
+```proto
+message BatchDeleteBooksRequest {
+  // The resource names of the books to delete.
+  // A maximum of 1000 books can be deleted in a batch.
+  repeated DeleteBookRequest requests = 1;
+}
+```
+
+- The request message **must** include one repeated field which accepts the
+  request messages specifying the resources to delete, as specified for
+  [standard Delete methods][request-message]. The field **should** be named
+  `requests`.
+- The request message **must not** contain any other required fields, and
+  **should not** contain other optional fields except those described in this or
+  another AIP.
+- The comment above the names field **should** document the maximum number of
+  requests allowed.
+- Filter-based matching **must not** be supported, because it makes it too easy for
```

Comment:
This probably deserves its own section. (It will need to be specifically cited frequently.)

Reply:
I can see that it might be cited frequently... but it's not some kind of additional functionality (like most sections are), it seems like a point pertinent to the request...

Reply:
Where would a filter come into the picture? BatchDelete takes an explicit set of resource names.

Reply:
It does, but API producers try to do it with a filter _all the time_ and we constantly have to say no.